[WIP] Updates to User Login and Invites

zkraime opened the pull request:
<h1>Purpose: </h1> Miscellaneous fixes and additions

<b> To do 
- [x] Custom email templates
- [x] Password length restriction
- [x] Fix UI errors
     - [x] Login view cutoff
     - [x] Restrict home page by authenticated (fix "Logged in as Sign Out" bug) 
     - [x] Add password change page for users
     - [x] Add forgot password link
     - [x] Display proper Django message class type (current 'patch' handles only success and error)
     - [ ] Handle warning/failure cases for send_email_invitation in Django-admin
- [x] Set up OSF email server for admin app
         - Needs to be tested on staging

Ghalko commented on admin/common_auth/forms.py:
```diff
@@ -15,3 +16,11 @@ def __init__(self, *args, **kwargs):
         self.fields['first_name'].required = True
         self.fields['last_name'].required = True
+
+    def clean_password1(self):
```
While it is certainly good to enforce password length, doesn't Django already do this somewhere?

zkraime replied:
@Ghalko Chris enlightened me on model-level character limits, so I will use that for password restriction instead of a form-level validation

Ghalko commented on admin/templates/base.html:
```diff
@@ -55,6 +55,9 @@
                 </a>
               </li>
               <li>
+                <a href="{% url 'password_change' %}" class="btn btn-flat">Change password</a>
```
I know this is probably not a good thing to skip, but we should probably have a different base page and user page where admin users can update their information.

So for now a button on the base page will be okay, but we should add a ticket for user settings.

zkraime replied:
@Ghalko I chatted with Chris about this briefly - that's exactly what I was thinking, having a user-settings page for updating account information. Same conclusion - leave it as is for now but keep it noted

Ghalko commented on admin/templates/common_auth/registration/invitation_email.html:
```diff
@@ -0,0 +1 @@
+{% extends 'registration/password_reset_email.html' %}
```
EOF.

Since some of these emails will be seen by more than just COS we should consider following the OSF in language used within the email.

zkraime replied:
@Ghalko I would absolutely agree. The current override is arbitrary (but not totally generic). I need to find an OSF template to model after

Ghalko commented on admin/templates/registration/password_reset_done.html:
```diff
@@ -0,0 +1,20 @@
+<div>
+<a href="{% url 'auth:login' %}">Return to home</a>
+</div>
+{% endblock %}
```
EOF

Ghalko commented on admin/templates/registration/password_reset_email.html:
```diff
@@ -0,0 +1,15 @@
+
+{% blocktrans %}-OSF Admin{% endblocktrans %}
+
+{% endautoescape %}
```
EOF

sloria commented:
Is there a JIRA ticket associated with this?

zkraime commented:
@sloria Actually no, I hadn't created one yet, I will do that

zkraime commented:
@cwisecarver Give this a quick pass? 
I have yet to test the emailing, but I asked Michael for the sendgrid password so hopefully I'll get that soon. 
The other thing is handling unsuccessful email invites on admin; I wanted to ask you about this real quick in person if you get the chance.
Thanks!

cwisecarver commented on admin/base/settings/defaults.py:
```diff
@@ -28,6 +29,14 @@
 # Email settings. Account created for testing. Password shouldn't be hardcoded
 EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
```
Just a note for devops, this should be set to `'django.core.mail.backends.smtp.EmailBackend'` in the > dev local.py.

cwisecarver commented on admin/base/urls.py:
```diff
@@ -16,8 +16,14 @@
-                            name='password_reset_complete'),
+                         url(r'^resetpassword/$', 'django.contrib.auth.views.password_reset', name='reset_password'),
+                         url(r'^password_reset_done/$', 'django.contrib.auth.views.password_reset_done', name='password_reset_done'),
+                         url(r'^accounts/password_change/$',
```
In the spirit of not hardcoding urls anywhere you don't have to I changed these lines to 
```python
                         url(r'^accounts/password_change/$',
                            'django.contrib.auth.views.password_change',
                            {'post_change_redirect': reverse_lazy('password_change_done')},
                            name="password_change"),
                         url(r'^accounts/password_change/done/$',
                            'django.contrib.auth.views.password_change_done', {'template_name': 'registration/password_change_done.html'}, name='password_change_done'),
                         )
```
And then imported `from django.core.urlresolvers import reverse_lazy` and everything still worked.

zkraime replied:
Thanks

cwisecarver commented on admin/common_auth/admin.py:
```diff
@@ -39,10 +39,6 @@ def send_email_invitation(self, request, queryset):
     send_email_invitation.short_description = 'Send email invitation to selected users'
 
     def save_model(self, request, obj, form, change):
```
You can just remove this method. You're not doing anything the grandfather class method isn't doing.

zkraime replied:
The save_model method? 
Yeah you're right

cwisecarver commented on admin/base/views.py:
```diff
@@ -5,14 +5,13 @@
     context = {'user': request.user}
     return render(request, 'home.html', context)
 
 def password_reset_done(request, **kwargs):
```
I don't think you're using this anywhere. At least not that pycharm find usages or the silver searcher could find.

zkraime replied:
The login_required decorator? Or the whole 'home' method?

cwisecarver replied:
It's been a while but I think I was talking about the password_reset_done method. Interestingly, it does seem to be used in the urls file. Not sure why the silver searcher didn't find it...

cwisecarver commented on admin/base/views.py:
```diff
@@ -5,14 +5,13 @@
+    messages.success(request, 'You have successfully reset your password and activated your admin account.')
     return redirect('auth:login')
 
 def password_reset_confirm_custom(request, **kwargs):
```
I know this method isn't part of this PR but I can't figure out what it does differently than [the regular one](https://docs.djangoproject.com/en/1.9/topics/auth/default/#django.contrib.auth.views.password_reset_confirm)

zkraime replied:
The only difference is that it is setting the confirmed flag to true

cwisecarver commented:
Done with this pass @Ghalko, @zkraime. Whoever picks this up let me know if there are questions.

Ghalko commented:
@zkraime What is the status of this?

zkraime commented:
@cwisecarver Please do a really quick pass over this to make sure I addressed your comments? 

Also, not sure how to go about handling warning and failure cases for email invites so I will leave that be for now since it's lower priority than other work (it's the one unchecked box)

cwisecarver commented:
Looks good to me @zkraime. Make sure you get with @mattclark when this goes to this being ready-to-merge about the email settings that will need to be put in local.py.

sloria commented:
As mentioned on the JIRA ticket: please update this branch with develop and resend the PR to develop. Thanks!
